### PR TITLE
Output external demangled procedures in the header file

### PR DESCRIPTION
**Pull request by ptomin**
They are sorted by (DLL name; class name; function name). External declarations from system or user-defined metafiles aren't here to avoid dublication.

### CODE REVIEW

File: src/Core/Program.cs [hunk 1]
```diff
@@ -374,6 +375,8 @@ public TypeLibraryDeserializer CreateTypeLibraryDeserializer()
         /// </summary>
         public Dictionary<string, Dictionary<FunctionType, IntrinsicProcedure>> Intrinsics { get; private set; }
 
+        public Dictionary<string, (string, string?, ExternalProcedure)> ExternalProcedures { get; private set; }
```
Comment (uxmal) [inline src/Core/Program.cs:378]: Please add a doc comment for this property.

File: src/Core/Program.cs [hunk 2]
```diff
@@ -822,5 +825,15 @@ public void AddGlobalField(Address address, DataType dt, string name)
             }
             fields.Add(offset, dt, name);
         }
+
```
Comment (uxmal) [inline src/Core/Program.cs:828]: In PE binaries, external procedures always are associated with a module name. On other platforms, like e.g. ELF, they are not. This signature should probably be changed to accept external procedures with `null` module names.

File: src/Decompiler/Decompiler.cs [hunk 3]
```diff
@@ -394,7 +394,7 @@ public void WriteGlobals(Program program, string filename, TextWriter w)
             w.WriteLine();
         }
     
-        public void WriteDecompiledTypes(Program program, string headerFilename, TextWriter w)
+        public void WriteDecompiledDecls(Program program, string headerFilename, TextWriter w)
```
Comment (uxmal) [inline src/Decompiler/Decompiler.cs:397]: Consider using the name `WriteDecompiledDeclarations` here.

File: src/Decompiler/Decompiler.cs [hunk 4]
```diff
@@ -414,6 +414,37 @@ public void WriteDecompiledTypes(Program program, string headerFilename, TextWri
...
+                p => (
+                    p.Value.Item1,
+                    p.Value.Item3.EnclosingType?.ToString(),
+                    p.Value.Item3.Name)
+            ).ToList();
+            if (externalProcedures.Count == 0)
+                return;
+            w.WriteLine("// Declarations for external procedures");
+            w.WriteLine();
+            foreach (var epKv in externalProcedures)
```
Comment (uxmal) [inline src/Decompiler/Decompiler.cs:433]: Suggestion: you can write the header of the `foreach` loop like this:
```C#
foreach (var (importName, (dllName, convention, ep)) in externalProcedures)
...
```

**Review by uxmal — COMMENTED** [review]
Nice work @ptomin. I've added some comments. The most significant one is about handling executable files that don't have module names associated with external procedures (e.g. ELF file format).

File: src/Decompiler/Decompiler.cs (hunk 4 above)
Comment (ptomin) [reply src/Decompiler/Decompiler.cs]: Done in 636607b

File: src/Decompiler/Decompiler.cs (hunk 3 above)
Comment (ptomin) [reply src/Decompiler/Decompiler.cs]: Done in a1e0d60

File: src/Core/Program.cs (hunk 2 above)
Comment (uxmal) [reply src/Core/Program.cs:834]: Also, let's add a doc comment here as well.

File: src/Core/Program.cs (hunk 1 above)
Comment (ptomin) [reply src/Core/Program.cs:384]: Done in 82b8276

File: src/Core/Program.cs (hunk 2 above)
Comment (ptomin) [reply src/Core/Program.cs:834]: Done in 4c6da22

**Comment by ptomin** [comment]
John, I have done requested changes.

**Comment by uxmal** [comment]
Nice contribution, Pavel!

**Pull request merged**